### ISSUE TITLE
Improve inferior process management

### ISSUE THREAD

**Issue by lionel-** [emacs-ess/ESS#416]
Creating this issue to collect all inferior management ideas.

#### List of inferior processes specific to frames

I think I'd have an easier time managing R processes if the inferior list was specific to frames. Then Emacs frames would behave a bit like RStudio projects (though they could still contain different packages, which I think is preferable and it would be hard to create a restriction anyway). @jennybc once [mentioned](https://github.com/krlmlr/here/issues/1#issuecomment-268871902) this as a "workflow nirvana" and I think the idea has grown on me ;)

There could be an option for whether to create frame-specific inferior lists, accepting one of those values: `'ask t nil`

**Comment by ess-core** [comment]
Hi Lionel:

I think this is a fantastic idea!  I have tried something like this
from time to time, but I always lapse back into a single *R*
buffer I guess because there is no convenient way to launch into this
habitually.

-- 
Rodney Sparapani, PhD
Assistant Professor of Biostatistics
Medical College of Wisconsin, Milwaukee

-----Original Message-----

**Comment by eddelbuettel** [comment]
I really don't quite understand what "frames" is supposed to mean here.  I just do `M-x R` a lot and then rename the buffer to say something more meaningful then *R:N* for various values of N .  That gives me pretty much exactly what Jenny's screenshot shows.  

But I probably just misunderstand something. Folks in the tibblyverse tend to have workflows I generally mostly shake my head about ;-)

**Comment by lionel-** [comment]
yes "frames" is what Emacs calls WM-level windows.

I can manually rename processes and switch. I just need something more automatic. Also a system where new buffers are never linked to an irrelevant process.

I use a buffer-oriented workflow as well, + tabs via elscreen, but can't think of something better than per-frame process lists to help me work with R (e.g. running revdeps for one package but be able to work on another package while they are running without ever having to deal with the process where revdep checks are performed).

**Comment by eddelbuettel** [comment]
Well, Emacs (running persistently in daemon mode) isn't the only tool on my computer.  

I tend to also work a lot in (also persistent) tmux (via [byobu](http://byobu.co/)) sessions so anything having to do with builds, compilations, tests, ... I tend to do there.   But I am probably much less of an Emacs power user than you.  It's only been about 23 years and I finally standardized by .emacs across machines a bit better.  Getting there.  Very slowly.  Doing what I can as a non-elisp writer.

**Comment by vspinu** [comment]
I don't think this is a good idea. This will lead to confusion as people use frames differently and all other emacs tools work across frames.  @jennybc opens one R studio instance per project. You can do the same with emacs. 

Also very few people know what process list is what it is for. It's 100% internal concept. In my opinion that process list is historical nuisance and ideally should be removed in the future.

But I would agree that process management is not ideal. I would suggest implementing a more project oriented system. For example, if an R session runs within a project then we automatically associate new buffers from that project to that R session. If there is no project then we consider directory of the R session and all subdirectories as part of the project. If there are several R sessions within same project then we look at the location of R scripts and associate scripts within same directory to same R session. 

As to the process names there is a system in place for automatic naming of processes `ess-gen-proc-buffer-name-function`. We can enhance those option if you think something is missing.

**Comment by lionel-** [comment]
Note that this would be completely optional, so wouldn't confuse anyone. Tying the process list to projects rather than frames might be a good option.

**Comment by lionel-** [comment]
The problem with projects is that there are use cases where your task involves several projects. For these cases, frame-based process lists might be better. I think it would make sense to support both.

**Comment by mmaechler** [comment]
Yes, I also work with several R processes  very very often.
And I do like the idea to allow (optional!) to have an R process to be "connected" to a project.
However, I would use the R processes often _not_ project oriented, but  R-version oriented: R 3.3.3,  R 3.4.0 beta,  R-devel  each having its own R process buffer,  and I have started them with `M-x R-devel`, `M-x R-3.3.3` etc since these are all in my path.
(and I want send R code to one of them, then same code to the next of them, etc: To see the difference, e.g. when chasing bugs).

@lionel- and "irrelevant process": Yes and no:  In any case I want to (keep) that an *.R script / package /... automatically connects to an R process, are at least semi-automatically: If there is one, ESS should keep showing (via IDO, I think) me the list of currently available R processes (+ the option "New") and if I type [Enter] it will choose the first of the list currently available ones.
If I open an R script I want to have R code executed almost instantly... and many of my R scripts will not belong to a "project".

**Comment by vspinu** [comment]
> and I want send R code to one of them, then same code to the next of them, etc: To see the difference, e.g. when chasing bugs

For such use you need to switch between R processes before every eval. 

If I understand correctly, Lionel is proposing that you keep same file open in separate frames and then eval the code from those buffers in different processes. I don't this working correctly? What if same process is open on two frames?   

> he problem with projects is that there are use cases where your task involves several projects. 

What evaluation pattern do you have in mind?

**Comment by mmaechler** [comment]
I did not want to oppose any of your proposals.  I just wanted to raise my use case as an important one for me as R developer... (and note that in such cases the R process can *not* be determined by an Rproj file).
Indeed have a frame-specific default R process seems like a very good idea to me.
